[BUG] Cannot pick a single value if that value is overflow/underflow

Currently, since using using Boost.Histogram's slicing as the backend, it has become impossible to "pick" an axis on an overflow or underflow bin. The following code fails:

```python
def test_pick_flowbin():
    w = 1e-2  # e.g. a cross section for a process
    x = np.random.normal(loc=0.4, scale=0.4, size=1000)
    y = np.random.normal(loc=0.6, scale=0.4, size=1000)

    h = bh.Histogram(
        bh.axis.Regular(3, 0, 1),
        bh.axis.Regular(3, 0, 1),
        storage=bh.storage.Weight(),
    )
    h.fill(x, y, weight=w)

    uf_slice = h[bh.tag.underflow, ...]
```

The reason is that it is picking is implemented via a workaround - a single bin is sliced:

```
--------------------------------------------- Captured log call --------------------------------------------------------
DEBUG    boost_histogram._internal.hist:hist.py:778
                Reduce with [reduce_command(slice_and_rebin(iaxis=0, begin=-1, end=0, merge=1, mode=slice_mode.crop))]
```

For a flow bin, this is an invalid slice. This would be much easier with a "pick" of some sort from Boost.Histogram. Possibly the only fix for now would be to special case under/overflow, and implement some ugly logic for it.

Discovered on Gitter by @andrzejnovak.

You are correct:

```python
@pytest.mark.parametrize(
    "ax",
    [bh.axis.Regular(3, 0, 1), bh.axis.Variable([0, 0.3, 0.6, 1])],
    ids=["regular", "variable"],
)
def test_pick_flowbin(ax):
    w = 1e-2  # e.g. a cross section for a process
    x = [-0.1, -0.1, 0.1, 0.1, 0.1]
    y = [-0.1, 0.1, -0.1, -0.1, 0.1]

    h = bh.Histogram(
        ax,
        ax,
        storage=bh.storage.Weight(),
    )
    h.fill(x, y, weight=w)

    uf_slice = h[bh.tag.underflow, ...]
    assert uf_slice.values(flow=True) == approx(np.array([1, 1, 0, 0, 0]) * w)

    uf_slice = h[..., bh.tag.underflow]
    assert uf_slice.values(flow=True) == approx(np.array([1, 2, 0, 0, 0]) * w)
```

```
>       reduced = self._hist.reduce(*slices)
E       ValueError: bins > 0 required

src/boost_histogram/_internal/hist.py:779: ValueError
----------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------
DEBUG    boost_histogram._internal.hist:hist.py:778 Reduce with [reduce_command(slice_and_rebin(iaxis=0, begin=-1, end=0, merge=1, mode=slice_mode.crop))]
==================================================================== short test summary info ====================================================================
FAILED tests/test_histogram_indexing.py::test_pick_flowbin[regular] - ValueError: bins > 0 required
=========================================================== 1 failed, 780 passed, 1 skipped in 2.85s ============================================================
```

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -2,7 +2,10 @@
 
 ## UPCOMING
 
-No changes yet.
+* Fix "picking" on a flow bin [#576][]
+
+[#576]: https://github.com/scikit-hep/boost-histogram/pull/576
+
 
 ## Version 1.0
 

--- a/src/boost_histogram/_internal/hist.py
+++ b/src/boost_histogram/_internal/hist.py
@@ -726,18 +726,22 @@ class Histogram:
             except RuntimeError:
                 pass
 
-        integrations = set()
-        slices = []
+        integrations: Set[int] = set()
+        slices: List[_core.algorithm.reduce_command] = []
+        pick_each: Dict[int, int] = dict()
 
         # Compute needed slices and projections
         for i, ind in enumerate(indexes):
             if hasattr(ind, "__index__"):
-                ind = slice(ind.__index__(), ind.__index__() + 1, sum)  # type: ignore
-
+                pick_each[i] = ind.__index__() + (  # type: ignore
+                    1 if self.axes[i].traits.underflow else 0
+                )
+                continue
             elif not isinstance(ind, slice):
                 raise IndexError(
                     "Must be a slice, an integer, or follow the locator protocol."
                 )
+
             # If the dictionary brackets are forgotten, it's easy to put a slice
             # into a slice - adding a nicer error message in that case
             if any(isinstance(v, slice) for v in (ind.start, ind.stop, ind.step)):
@@ -778,15 +782,25 @@ class Histogram:
         logger.debug("Reduce with %s", slices)
         reduced = self._hist.reduce(*slices)
 
-        if not integrations:
-            return self._new_hist(reduced)
-        projections = [i for i in range(self.ndim) if i not in integrations]
+        if pick_each:
+            my_slice = tuple(
+                pick_each.get(i, slice(None)) for i in range(reduced.rank())
+            )
+            logger.debug("Slices: %s", my_slice)
+            axes = [
+                reduced.axis(i) for i in range(reduced.rank()) if i not in pick_each
+            ]
+            logger.debug("Axes: %s", axes)
+            new_reduced = reduced.__class__(axes)
+            new_reduced.view(flow=True)[...] = reduced.view(flow=True)[my_slice]
+            reduced = new_reduced
+            integrations = {i - sum(j <= i for j in pick_each) for i in integrations}
 
-        return (
-            self._new_hist(reduced.project(*projections))
-            if projections
-            else reduced.sum(flow=True)
-        )
+        if integrations:
+            projections = [i for i in range(reduced.rank()) if i not in integrations]
+            reduced = reduced.project(*projections)
+
+        return self._new_hist(reduced) if reduced.rank() > 0 else reduced.sum(flow=True)
 
     def __setitem__(
         self, index: IndexingExpr, value: Union[ArrayLike, Accumulator]

--- a/tests/test_histogram_indexing.py
+++ b/tests/test_histogram_indexing.py
@@ -1,6 +1,7 @@
 import numpy as np
 import pytest
 from numpy.testing import assert_array_equal
+from pytest import approx
 
 import boost_histogram as bh
 
@@ -349,6 +350,30 @@ def test_pick_int_category():
     assert_array_equal(h[{2: bh.loc(3)}].view(), vals)
     assert_array_equal(h[:, :, bh.loc(5)].view(), vals + 1)
     assert_array_equal(h[:, :, bh.loc(7)].view(), 0)
+
+
+@pytest.mark.parametrize(
+    "ax",
+    [bh.axis.Regular(3, 0, 1), bh.axis.Variable([0, 0.3, 0.6, 1])],
+    ids=["regular", "variable"],
+)
+def test_pick_flowbin(ax):
+    w = 1e-2  # e.g. a cross section for a process
+    x = [-0.1, -0.1, 0.1, 0.1, 0.1]
+    y = [-0.1, 0.1, -0.1, -0.1, 0.1]
+
+    h = bh.Histogram(
+        ax,
+        ax,
+        storage=bh.storage.Weight(),
+    )
+    h.fill(x, y, weight=w)
+
+    uf_slice = h[bh.tag.underflow, ...]
+    assert uf_slice.values(flow=True) == approx(np.array([1, 1, 0, 0, 0]) * w)
+
+    uf_slice = h[..., bh.tag.underflow]
+    assert uf_slice.values(flow=True) == approx(np.array([1, 2, 0, 0, 0]) * w)
 
 
 def test_axes_tuple():

